druntime/src/core/internal/convert.d

```diff
@@ -768,54 +770,50 @@ nothrow pure @safe unittest
+    return ctfe_alloc(T.sizeof);
+}
+
 @trusted pure nothrow @nogc
 const(ubyte)[] toUbyte(T)(const ref T val) if (is(T == delegate) || is(T : V*, V) && __traits(getAliasThis, T).length == 0)
 {
     pragma(inline, true);
-    if (__ctfe)
-    {
-        if (val !is null) assert(0, "Unable to compute byte representation of non-null pointer at compile time");
-        return ctfe_alloc(T.sizeof);
-    }
-    else
+    return __ctfe ? toUbyte_delegate_ctfe(val)
+                  : (cast(const(ubyte)*)&val)[0 .. T.sizeof];
+}
+
+private const(ubyte)[] toUbyte_aggregate_ctfe(T)(const return ref scope T val)
+{
+    pragma(inline, false);
+    ubyte[] bytes = ctfe_alloc(T.sizeof);
+    foreach (key, ref cur; val.tupleof)
     {
-        return (cast(const(ubyte)*)&val)[0 .. T.sizeof];
+        static if (is(typeof(cur) EType == enum)) // Odd style is to avoid template instantiation in most cases.
+            alias CurType = OriginalType!EType;
+        else
+            alias CurType = typeof(cur);
+        static if (is(CurType == struct) || is(CurType == union) || __traits(isStaticArray, CurType) || !is(typeof(cur is null)))
+        {
+            bytes[val.tupleof[key].offsetof .. val.tupleof[key].offsetof + CurType.sizeof] = toUbyte(cur)[];
+        }
+        else
+        {
+            assert(cur is null, "Unable to compute byte representation of non-null reference field at compile time");
+            //skip, because val bytes are zeros
+        }
     }
+    return bytes;
 }
 
 @trusted pure nothrow @nogc
 const(ubyte)[] toUbyte(T)(const return ref scope T val) if (is(T == struct) || is(T == union))
 {
     pragma(inline, true);
-    if (__ctfe)
-    {
-        ubyte[] bytes = ctfe_alloc(T.sizeof);
-        foreach (key, ref cur; val.tupleof)
-        {
-            static if (is(typeof(cur) EType == enum)) // Odd style is to avoid template instantiation in most cases.
-                alias CurType = OriginalType!EType;
-            else
-                alias CurType = typeof(cur);
-            static if (is(CurType == struct) || is(CurType == union) || __traits(isStaticArray, CurType) || !is(typeof(cur is null)))
-            {
-                bytes[val.tupleof[key].offsetof .. val.tupleof[key].offsetof + CurType.sizeof] = toUbyte(cur)[];
-            }
-            else
-            {
-                assert(cur is null, "Unable to compute byte representation of non-null reference field at compile time");
-                //skip, because val bytes are zeros
-            }
-        }
-        return bytes;
-    }
-    else
-    {
-        // We're escaping a reference to `val` here because we cannot express
-        // ref return + scope, it's currently seen as ref + return scope
-        // https://issues.dlang.org/show_bug.cgi?id=22541
-        // Once fixed, the @system lambda should be removed
-        return (() @system => (cast(const(ubyte)*)&val)[0 .. T.sizeof])();
```

Comment:
As the mentioned bug was fixed in 2022, I removed the lambda here, it prevented inlining, too.